### ISSUE TITLE
[Feature]: Let merge this :) 

### ISSUE THREAD

**Issue by chipweinberger** [tnc1997/flutter-blue-plus#1]
### What is your feature request?

I'm looking it over, and i'm impressed with how it turned out. Thanks for taking the initiative.

**Comment by chipweinberger** [comment]
I added you as a collaborator. Could you handle the work of merging the remaining commits?

Then when you are ready, do a "Release 1.35.0" commit & add a git tag.

**Comment by chipweinberger** [comment]
I just added this setting to `pub.dev`, so hopefully just by pushing a tag of the format `X.X.X`, it will publish to pub.dev.

<img width="568" alt="Screenshot 2024-12-23 at 11 16 56 AM" src="https://github.com/user-attachments/assets/0ba4d02e-ee1f-49c1-967c-381ad6d153e7" />

**Comment by chipweinberger** [comment]
Also just a bit of context - the last time we talked about all this I was right before product launch, so I very much wanted to avoid active development on the main branch.

But now the launch already happened and the launch bugs are ironed out. So now is a much better time for these developments.

**Comment by tnc1997** [comment]
Hi @chipweinberger, it's great that you're keen to merge these platform interface changes into the upstream repository!

In order to minimise your time commitment would you like me to maintain the Linux and Web platforms separately?

**Comment by chipweinberger** [comment]
It's a good question. 

I think just all one repository is fine. But up to you!

**Comment by tnc1997** [comment]
> I think just all one repository is fine.

That sounds good to me, if you're happy with all platform implementations in the upstream repository, then I will open a Pull Request to merge the changes in this repository.

Have you got any upcoming changes that I should plan to merge into this fork before opening the Pull Request?

**Comment by chipweinberger** [comment]
no need for a pr! / you can merge as you please

you could work off master branch too if you want!

**Comment by chipweinberger** [comment]
as for upcoming changes, none. But these eventually will be fixed:

https://github.com/chipweinberger/flutter_blue_plus/issues/1047

**Comment by tnc1997** [comment]
Thank you for confirming that, I have prepared a branch that we should be able to [merge](https://github.com/chipweinberger/flutter_blue_plus/pull/1091) into upstream's master.

Once merged, we should publish the [platform interface](https://github.com/tnc1997/flutter-blue-plus/tree/feature/platform-interface/packages/flutter_blue_plus_platform_interface), then the implementations, then the [main package](https://github.com/tnc1997/flutter-blue-plus/tree/feature/platform-interface/packages/flutter_blue_plus) itself.

> Are you okay to handle the work of merging the remaining commits?

I am more than happy to handle the merging of the linked Pull Request into master if you are happy for me to.

> I added this setting to pub.dev, so hopefully just by pushing a tag of the format X.X.X, we can publish to pub.dev.

Just to double check, does this take into account that we will have a monorepo, that contains multiple packages?

**Comment by chipweinberger** [comment]
> Just to double check, does this take into account that we will have a monorepo, that contains multiple packages?

not sure

**Comment by tnc1997** [comment]
> I added this setting to `pub.dev`, so hopefully just by pushing a tag of the format `X.X.X`, we can publish to pub.dev.

Would you be able to update the tag that is used in the automated publishing settings to handle multiple packages?

> If your repository contains multiple packages, give each a separate tag-pattern. Consider using a tag-pattern like my_package_name-v{{version}} for a package named my_package_name.

Based on the guidance a tag pattern of `flutter_blue_plus-v{{version}}` for the main package would be good to use.

**Comment by chipweinberger** [comment]
okay

<img width="576" alt="Screenshot 2025-01-05 at 11 44 09 AM" src="https://github.com/user-attachments/assets/27aaf3ca-5c79-4bd5-a3b8-62dd448264ab" />

**Comment by tnc1997** [comment]
Looks good to me, I will merge the associated Pull Request, then set up some GitHub Actions to publish the packages.

Upon reviewing the documentation, the initial version of the newly added packages will need to published manually.

**Comment by chipweinberger** [comment]
what is the current state of master branch on `flutter_blue_plus`?

please email me so we can exchange contact details :)

**Comment by tnc1997** [comment]
> what is the current state of master branch on `flutter_blue_plus`?

The `master` branch has the platform interface and the federated packages ready for reviewing, testing, publishing.

**Comment by chipweinberger** [comment]
okay, i pushed all the packages.

**Comment by chipweinberger** [comment]
I've also added all these actions. So hopefully you can figure out how to publish easily.

<img width="798" alt="Screenshot 2025-01-14 at 11 40 58 PM" src="https://github.com/user-attachments/assets/6947e12b-57a8-4d22-97e4-7724b25ec2b2" />

**Comment by chipweinberger** [comment]
last, just a small style thing, please push a "Release 1.35.0" commit. Just so we can easily see each version in the commit history.

**Comment by chipweinberger** [comment]
<img width="328" alt="Screenshot 2025-01-14 at 11 47 13 PM" src="https://github.com/user-attachments/assets/4ba37e98-5c27-4630-9ceb-5946fd55ce8a" />

**Comment by tnc1997** [comment]
> So hopefully you can figure out how to publish easily.

Now that the initial versions have been published we should be able to push git tag(s) to run the relevant workflow(s).

**Comment by tnc1997** [comment]
> please push a "Release 1.35.0" commit

Out of interest would you like to differentiate between the different packages (and versions) in the commit message?

**Comment by chipweinberger** [comment]
Yes, we probably ought to. I'm not quite sure, but whatever makes the most sense.

**Comment by chipweinberger** [comment]
maybe it's easier to use the same version number for all packages. 

bascially, treat the version number as a increasing 'timestamp'. Whenever we release a package, we use the latest version number.

For example:
1. update linux to 1.35.1
2. release linux 1.35.1 (other packages stay on 1.35.0)
3. update linux to 1.35.2
4. release linux 1.35.2 (other packages stay the 1.35.0)
5. update linux & macos to 1.35.3
6. release linux & macos to 1.35.3 (other packages stay the 1.35.0)

**Comment by chipweinberger** [comment]
Okay, I think I know how to do it.

We will use a single version number, essentially.

Whenever we update *any* package, we'll update `flutter_blue_plus/pubspec.yaml` to the latest version, and then always release a new `flutter_blue_plus` package + whatever other package changed.

```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_blue_plus_android: ^1.35.0
  flutter_blue_plus_ios: ^1.35.0
  flutter_blue_plus_linux: ^1.0.0
  flutter_blue_plus_macos: ^1.35.0
  flutter_blue_plus_platform_interface: ^1.0.0
  flutter_blue_plus_web: ^1.0.0
```

Also, we should remove the `^` symbol. I dont want people to mix and match versions, it just creates unnecessary complexity. For every `flutter_blue_plus`, it should have a specific set of versions that it is meant for.

**Comment by chipweinberger** [comment]
FBP has had a single version number for years. 

IMO, let's start with one version number. We can always introduce ranges later.

**Comment by tnc1997** [comment]
> IMO, let's start with one version number. It's simpler for us, and simpler for users.

Of course, it is your decision as the lead maintainer, I was simply pointing out what the other Flutter packages do.

**Comment by chipweinberger** [comment]
Lead maintainer, for now :)

**Comment by tnc1997** [comment]
> Thanks again for all your work on platform interface.

No problem at all, it is my pleasure!

I am very grateful to be able to use `flutter_blue_plus` on more platforms.

Thank you very much for your support reviewing, testing, publishing the platform interface and federated packages.

**Comment by chipweinberger** [comment]
@https://github.com/ekuleshov, just want to make you aware of this: https://github.com/tnc1997/flutter-blue-plus/issues/1#issuecomment-2560190867

you can push tags and it should release to pub.dev, assuming we setup the github actions correctly.

In terms of version numbers, I'm flexible. I prefer using `1.X.X` for the main FBP repo. For the others, I don't really care :)

**Comment by chipweinberger** [comment]
@tnc1997 , what is the purpose of this commit? https://github.com/chipweinberger/flutter_blue_plus/commit/3b5e4d186d1734b1eb9e3c0fc092fbb727605e47

Why do they all need to return bool?

**Comment by chipweinberger** [comment]
Also worth pointing out, the interface will change dramatically when we implement this: https://github.com/chipweinberger/flutter_blue_plus/issues/795

**Comment by ekuleshov** [comment]
> Also worth pointing out, the interface will change dramatically when we implement this: [chipweinberger#795](https://github.com/chipweinberger/flutter_blue_plus/issues/795)

Not necessarily. E.g. if we'd just rely on the order of services/chars in the public api. Then internally could use indexes or use internal subclass of Guid with an index

**Comment by tnc1997** [comment]
> Why do they all need to return bool?

After reviewing the Android/iOS/macOS platform implementations, this is the result that those method channels return, so returning a `bool` aligns the platform interface with these pre-existing federated platform implementations. If you like we could update the Android/iOS/macOS platform implementations as part of this piece of work to not return a `bool`.

**Comment by tnc1997** [comment]
> In terms of version numbers, I'm flexible. I prefer using `1.X.X` for the main `flutter_blue_plus` repo.

That sounds good, would you be comfortable for us to use semantic versioning for `flutter_blue_plus_android` `flutter_blue_plus_darwin` `flutter_blue_plus_linux` `flutter_blue_plus_platform_interface` `flutter_blue_plus_web`, but continue to use `1.X.X` for the main `flutter_blue_plus` package that developers would add as a dependency?
